### DISCUSSION TITLE
Help me on Unicode UTF-8

### DISCUSSION THREAD

**Discussion by kydang789** [HASwitchPlate/openHASP#258] [Q&A] [answered]
## Questions

I make a label in Unicode : "Chi**ế**u **đ**iểm" actually I got just "Chi**ê**u i**ê**m"

![image](https://user-images.githubusercontent.com/68098745/146635180-44d2a1b5-6eb9-44bd-a812-d90286f0d3c7.png)

Pls help me how to fix

We only keep bug reports and active requests being worked on in the issues tracker.

For general questions we have opened the [discussion forum](https://github.com/HASwitchPlate/openHASP/discussions) on github and [#openHASP channel](https://discord.gg/VCWyuhF) on Discord.

If you are not sure if your question is related to the code development, feel free to ask!

**Comment by fvanroie** [comment]
I suppose these letters are from the Vietnamese language?
The Vietnamese [character set](https://haswitchplate.github.io/openHASP-docs/latest/design/fonts/#character-sets) is not yet supported, but we can add it if needed.

- First we need to create a list of required characters and add them to `src\font\encodings.json`
- We can then generate the LVGL font
- Compile openHASP witht the Vietnamese font encoding

I'm not sure if all glyphs are available in Roboto Condensed, so you will need to test it.

**Reply by nagyrobi** [reply]
One more reason for openhasplite.ttf...

**Reply by kydang789** [reply]
I shall try it

**Comment by fvanroie** [comment]
I have added the following [Vietnamese characters](https://haswitchplate.github.io/openHASP-docs/latest/design/fonts/#vietnamese) to [`encodings.json`](https://github.com/HASwitchPlate/openHASP/commit/0ab05e6b4a7e9e624645a6ab2e92f74d9e986d26):

> Á À Â Ă Ã Ấ Ầ Ắ Ằ Ẫ Ẵ Ả Ẩ Ẳ Ạ Ậ Ặ Đ É È Ê Ẽ Ế Ề Ễ Ẻ Ể Ẹ Ệ Í Ì Ĩ Ỉ Ị Ó Ò Ô Õ Ố Ồ Ỗ Ỏ Ơ Ổ Ọ Ớ Ờ Ỡ Ộ Ở Ợ Ú Ù Ũ Ủ Ư Ụ Ứ Ừ Ữ Ử Ự Ý Ỳ Ỹ Ỷ Ỵ á à â ă ã ấ ầ ắ ằ ẫ ẵ ả ẩ ẳ ạ ậ ặ đ é è ê ẽ ế ề ễ ẻ ể ẹ ệ í ì ĩ ỉ ị ó ò ô õ ố ồ ỗ ỏ ơ ổ ọ ớ ờ ỡ ộ ở ợ ú ù ũ ủ ư ụ ứ ừ ữ ử ự ý ỳ ỹ ỷ ỵ ₫ 

It is important to encode the accents and characters into **one** single codepoint for LVGL to render the correct letter. This can be achieved by converting UTF-8 text to **[Unicode NFC format](https://dencode.com/string/unicode-normalization?v=Chi%C3%AA%CC%81u%20%C4%91i%C3%AA%CC%89m)**, NFC means the characters are decomposed and then re-composed by **canonical equivalence**.

This conversion needs to be done in advance, otherwise ArduinoJSON or LVGL will treat the character and accents separately...

For [example](https://onlineutf8tools.com/convert-utf8-to-code-points?input=Chi%E1%BA%BFu%20%C4%91i%E1%BB%83m%0A&base=hexadecimal&custom-base=16&separator=%20):
- `ế` must be sent as the single codepoint [`U+01EBF`](http://www.isthisthingon.org/unicode/index.phtml?glyph=1EBF) and not separate characters `e` `^` `´`.
- `ể` must be sent as the single codepoint [`U+01EC3`](http://www.isthisthingon.org/unicode/index.phtml?glyph=01EC3) and not separate characters `e` `^` `'`.

Then `Chiếu điểm` looks like:

![image](https://user-images.githubusercontent.com/15969459/148709604-1de1dbbe-62d9-4659-b681-876b4f1ae626.png)

**Comment by fvanroie** [comment, marked as answer]
You need to add this code to `include\user_config_override.h` for the Vietnamese font to be used:

```cpp
#define HASP_FONT_1 robotocondensed_regular_16_vietnamese // Use available fonts from src/fonts directory
#define HASP_FONT_2 robotocondensed_regular_24_vietnamese
#define HASP_FONT_3 robotocondensed_regular_32_vietnamese
#define HASP_FONT_4 robotocondensed_regular_38_vietnamese

#define ROBOTOCONDENSED_REGULAR_16_VIETNAMESE 1 // Enable the selected fonts
#define ROBOTOCONDENSED_REGULAR_24_VIETNAMESE 1
#define ROBOTOCONDENSED_REGULAR_32_VIETNAMESE 1
#define ROBOTOCONDENSED_REGULAR_38_VIETNAMESE 1

#define HASP_FONT_SIZE_1 16 // 12 orig              // Define used font sizes
#define HASP_FONT_SIZE_2 24 // 16 orig
#define HASP_FONT_SIZE_3 32 // 24 orig
#define HASP_FONT_SIZE_4 38 // 32 orig
```

**Comment by fvanroie** [comment]
This has been tested and works:
![image](https://user-images.githubusercontent.com/15969459/151684668-8cb18524-3a99-4472-a84e-d67693518b29.png)

**Reply by erwin** [reply]
@fvanroie Same question but for Chinese? I did check and it looks like `Roboto Condensed` does not contain any Chinese characters...

Just to test it, I did send some messages including Chinese characters to the display, but they were just ignored.

Would I need to extract the exact characters that I would be using, as most Chinese fonts have 2+ orders of magnitude more characters than US only fonts...

I'm using openHASP on the Lanbon L8 with Home Assistant

**Reply by fvanroie** [reply]
Indeed we need to select the codepoints/characters to include in the firmware. We will also need to find a suitable CJK font.
In LVGL they use `SimSun.woff` with these [1269 characters](https://github.com/lvgl/lvgl/blob/0b5a1d4b23975b920ff841ea9cd038802f51711b/scripts/built_in_font/generate_all.py#L77-L78).

I think this is Simplified Chinese... Will these be enough? Is there a list of most common characters? Maybe like from the HSK Levels?
It could be that there is only room for 1 default font size instead of 4 font sizes.

[SimSun](https://www.fontke.com/search/simsun.woff/) is a Serif font, but I would rater use a Sans-Serif one instead, like [Noto Sans Simplified Chinese](https://fonts.google.com/noto/specimen/Noto+Sans+SC).

**Reply by erwin** [reply]
> 盗提陽帯鼻画輕ッ冊ェル写父ぁフ結想正四O夫源庭場天續鳥れ講猿苦階給了製守

This list of characters could be Japanese because it has Hiragana and Katana characters mixed in, along with traditional Chinese characters. Could also be a dump of GBK as explained below.

Katakana: ェ ィ ッ ミ テ ラ ス ゲ
Hiragana: ぁ れ ぺ ァ コ っ よ ヤ ジ く つ ん ゃ て お が う

It's very tricky to try and pick a subset of Chinese Characters

To figure out if a set of characters is simplified or traditional, I think the easiest way is to look for Airport (机场 simplified vs 機場 traditional) or Square/Marketplace (广场 simplified vs 廣場 traditional) as these are very common words and will show up on almost any word list.

The Chinese Standard "GB 2312" includes 6763 characters, including:
- Level One Chinese Characters: 3755
- Level Two Chinese Characters: 3008
- Additional 682 Characters of:
    - Latin Characters
    - Greek Characters
    - Japanese Hiragana and Katakana
    - Cyrillic Characters

The Characters represented in "GB 2312" Level One and Level Two cover 99.75% of the character use in Mainland China.

To account for many [Chinese] people's names, and ancient Chinese writing (still a required school subject) GB 2312 is insufficient and GBK or GB 18030 are used instead. I've used phones in China that only support GB 2312 and it's painful, since many common names can't be saved accurately in the phone's address book...

That said, Beijing Language University has an official list of the most common 3500 characters:

https://faculty.blcu.edu.cn/system/resource/tsites/download.jsp?ap=jswzfj&fl=%2F_vsl%2F590FCD1D8620A87AD77CAA10752FF6C5%2FAC495CEA%2F24E00&tid=25913&fid=1043

For embedded systems, I was able to find several articles of developers using the most common 2500 characters. Trying to get down to the most common 1000 characters or something will get very tricky though - quite a lot of useful characters would be tossed out.

Most common 2500 characters. [Source](https://www.cnblogs.com/skyivben/archive/2012/12/13/2815944.html)

```
一乙
二十丁厂七卜人入八九几儿了力乃刀又
三于干亏士工土才寸下大丈与万上小口巾山千乞川亿个勺久凡及夕丸么广亡门义之尸弓己已子卫也女飞刃习叉马乡
丰王井开夫天无元专云扎艺木五支厅不太犬区历尤友匹车巨牙屯比互切瓦止少日中冈贝内水见午牛手毛气升长仁什片仆化仇币仍仅斤爪反介父从今凶分乏公仓月氏勿欠风丹匀乌凤勾文六方火为斗忆订计户认心尺引丑巴孔队办以允予劝双书幻
玉刊示末未击打巧正扑扒功扔去甘世古节本术可丙左厉右石布龙平灭轧东卡北占业旧帅归且旦目叶甲申叮电号田由史只央兄叼叫另叨叹四生失禾丘付仗代仙们仪白仔他斥瓜乎丛令用甩印乐句匆册犯外处冬鸟务包饥主市立闪兰半汁汇头汉宁穴它讨写让礼训必议讯记永司尼民出辽奶奴加召皮边发孕圣对台矛纠母幼丝
式刑动扛寺吉扣考托老执巩圾扩扫地扬场耳共芒亚芝朽朴机权过臣再协西压厌在有百存而页匠夸夺灰达列死成夹轨邪划迈毕至此贞师尘尖劣光当早吐吓虫曲团同吊吃因吸吗屿帆岁回岂刚则肉网年朱先丢舌竹迁乔伟传乒乓休伍伏优伐延件任伤价份华仰仿伙伪自血向似后行舟全会杀合兆企众爷伞创肌朵杂危旬旨负各名多争色壮冲冰庄庆亦刘齐交次衣产决充妄闭问闯羊并关米灯州汗污江池汤忙兴宇守宅字安讲军许论农讽设访寻那迅尽导异孙阵阳收阶阴防奸如妇好她妈戏羽观欢买红纤级约纪驰巡
寿弄麦形进戒吞远违运扶抚坛技坏扰拒找批扯址走抄坝贡攻赤折抓扮抢孝均抛投坟抗坑坊抖护壳志扭块声把报却劫芽花芹芬苍芳严芦劳克苏杆杠杜材村杏极李杨求更束豆两丽医辰励否还歼来连步坚旱盯呈时吴助县里呆园旷围呀吨足邮男困吵串员听吩吹呜吧吼别岗帐财针钉告我乱利秃秀私每兵估体何但伸作伯伶佣低你住位伴身皂佛近彻役返余希坐谷妥含邻岔肝肚肠龟免狂犹角删条卵岛迎饭饮系言冻状亩况床库疗应冷这序辛弃冶忘闲间闷判灶灿弟汪沙汽沃泛沟没沈沉怀忧快完宋宏牢究穷灾良证启评补初社识诉诊词译君灵即层尿尾迟局改张忌际陆阿陈阻附妙妖妨努忍劲鸡驱纯纱纳纲驳纵纷纸纹纺驴纽
奉玩环武青责现表规抹拢拔拣担坦押抽拐拖拍者顶拆拥抵拘势抱垃拉拦拌幸招坡披拨择抬其取苦若茂苹苗英范直茄茎茅林枝杯柜析板松枪构杰述枕丧或画卧事刺枣雨卖矿码厕奔奇奋态欧垄妻轰顷转斩轮软到非叔肯齿些虎虏肾贤尚旺具果味昆国昌畅明易昂典固忠咐呼鸣咏呢岸岩帖罗帜岭凯败贩购图钓制知垂牧物乖刮秆和季委佳侍供使例版侄侦侧凭侨佩货依的迫质欣征往爬彼径所舍金命斧爸采受乳贪念贫肤肺肢肿胀朋股肥服胁周昏鱼兔狐忽狗备饰饱饲变京享店夜庙府底剂郊废净盲放刻育闸闹郑券卷单炒炊炕炎炉沫浅法泄河沾泪油泊沿泡注泻泳泥沸波泼泽治怖性怕怜怪学宝宗定宜审宙官空帘实试郎诗肩房诚衬衫视话诞询该详建肃录隶居届刷屈弦承孟孤陕降限妹姑姐姓始驾参艰线练组细驶织终驻驼绍经贯
奏春帮珍玻毒型挂封持项垮挎城挠政赴赵挡挺括拴拾挑指垫挣挤拼挖按挥挪某甚革荐巷带草茧茶荒茫荡荣故胡南药标枯柄栋相查柏柳柱柿栏树要咸威歪研砖厘厚砌砍面耐耍牵残殃轻鸦皆背战点临览竖省削尝是盼眨哄显哑冒映星昨畏趴胃贵界虹虾蚁思蚂虽品咽骂哗咱响哈咬咳哪炭峡罚贱贴骨钞钟钢钥钩卸缸拜看矩怎牲选适秒香种秋科重复竿段便俩贷顺修保促侮俭俗俘信皇泉鬼侵追俊盾待律很须叙剑逃食盆胆胜胞胖脉勉狭狮独狡狱狠贸怨急饶蚀饺饼弯将奖哀亭亮度迹庭疮疯疫疤姿亲音帝施闻阀阁差养美姜叛送类迷前首逆总炼炸炮烂剃洁洪洒浇浊洞测洗活派洽染济洋洲浑浓津恒恢恰恼恨举觉宣室宫宪突穿窃客冠语扁袄祖神祝误诱说诵垦退既屋昼费陡眉孩除险院娃姥姨姻娇怒架贺盈勇怠柔垒绑绒结绕骄绘给络骆绝绞统
耕耗艳泰珠班素蚕顽盏匪捞栽捕振载赶起盐捎捏埋捉捆捐损都哲逝捡换挽热恐壶挨耻耽恭莲莫荷获晋恶真框桂档桐株桥桃格校核样根索哥速逗栗配翅辱唇夏础破原套逐烈殊顾轿较顿毙致柴桌虑监紧党晒眠晓鸭晃晌晕蚊哨哭恩唤啊唉罢峰圆贼贿钱钳钻铁铃铅缺氧特牺造乘敌秤租积秧秩称秘透笔笑笋债借值倚倾倒倘俱倡候俯倍倦健臭射躬息徒徐舰舱般航途拿爹爱颂翁脆脂胸胳脏胶脑狸狼逢留皱饿恋桨浆衰高席准座脊症病疾疼疲效离唐资凉站剖竞部旁旅畜阅羞瓶拳粉料益兼烤烘烦烧烛烟递涛浙涝酒涉消浩海涂浴浮流润浪浸涨烫涌悟悄悔悦害宽家宵宴宾窄容宰案请朗诸读扇袜袖袍被祥课谁调冤谅谈谊剥恳展剧屑弱陵陶陷陪娱娘通能难预桑绢绣验继
球理捧堵描域掩捷排掉堆推掀授教掏掠培接控探据掘职基著勒黄萌萝菌菜萄菊萍菠营械梦梢梅检梳梯桶救副票戚爽聋袭盛雪辅辆虚雀堂常匙晨睁眯眼悬野啦晚啄距跃略蛇累唱患唯崖崭崇圈铜铲银甜梨犁移笨笼笛符第敏做袋悠偿偶偷您售停偏假得衔盘船斜盒鸽悉欲彩领脚脖脸脱象够猜猪猎猫猛馅馆凑减毫麻痒痕廊康庸鹿盗章竟商族旋望率着盖粘粗粒断剪兽清添淋淹渠渐混渔淘液淡深婆梁渗情惜惭悼惧惕惊惨惯寇寄宿窑密谋谎祸谜逮敢屠弹随蛋隆隐婚婶颈绩绪续骑绳维绵绸绿
琴斑替款堪搭塔越趁趋超提堤博揭喜插揪搜煮援裁搁搂搅握揉斯期欺联散惹葬葛董葡敬葱落朝辜葵棒棋植森椅椒棵棍棉棚棕惠惑逼厨厦硬确雁殖裂雄暂雅辈悲紫辉敞赏掌晴暑最量喷晶喇遇喊景践跌跑遗蛙蛛蜓喝喂喘喉幅帽赌赔黑铸铺链销锁锄锅锈锋锐短智毯鹅剩稍程稀税筐等筑策筛筒答筋筝傲傅牌堡集焦傍储奥街惩御循艇舒番释禽腊脾腔鲁猾猴然馋装蛮就痛童阔善羡普粪尊道曾焰港湖渣湿温渴滑湾渡游滋溉愤慌惰愧愉慨割寒富窜窝窗遍裕裤裙谢谣谦属屡强粥疏隔隙絮嫂登缎缓编骗缘
瑞魂肆摄摸填搏塌鼓摆携搬摇搞塘摊蒜勤鹊蓝墓幕蓬蓄蒙蒸献禁楚想槐榆楼概赖酬感碍碑碎碰碗碌雷零雾雹输督龄鉴睛睡睬鄙愚暖盟歇暗照跨跳跪路跟遣蛾蜂嗓置罪罩错锡锣锤锦键锯矮辞稠愁筹签简毁舅鼠催傻像躲微愈遥腰腥腹腾腿触解酱痰廉新韵意粮数煎塑慈煤煌满漠源滤滥滔溪溜滚滨粱滩慎誉塞谨福群殿辟障嫌嫁叠缝缠
静碧璃墙撇嘉摧截誓境摘摔聚蔽慕暮蔑模榴榜榨歌遭酷酿酸磁愿需弊裳颗嗽蜻蜡蝇蜘赚锹锻舞稳算箩管僚鼻魄貌膜膊膀鲜疑馒裹敲豪膏遮腐瘦辣竭端旗精歉熄熔漆漂漫滴演漏慢寨赛察蜜谱嫩翠熊凳骡缩
慧撕撒趣趟撑播撞撤增聪鞋蕉蔬横槽樱橡飘醋醉震霉瞒题暴瞎影踢踏踩踪蝶蝴嘱墨镇靠稻黎稿稼箱箭篇僵躺僻德艘膝膛熟摩颜毅糊遵潜潮懂额慰劈
操燕薯薪薄颠橘整融醒餐嘴蹄器赠默镜赞篮邀衡膨雕磨凝辨辩糖糕燃澡激懒壁避缴
戴擦鞠藏霜霞瞧蹈螺穗繁辫赢糟糠燥臂翼骤
鞭覆蹦镰翻鹰
警攀蹲颤瓣爆疆
壤耀躁嚼嚷籍魔灌
蠢霸露
囊
罐
```

**Reply by nagyrobi** [reply]
Wow.

**Reply by fvanroie** [reply]
It can work, but I suggest to use PSram only for these fonts because they are very big. You will need 8MB flash and 4MB PSram.
For example, these fonts contain 3500 characters with MDI icons, but without ASCII:
```sh
    296kB  notosans16.bin
    518kB  notosans24.bin
    746kB  notosans32.bin
  1.004kB  notosans40.bin
  1.240kB  notosans48.bin
```
You can use a jsonl like this:
```json
{"page":9,"comment":"---------- Page 9 ----------"}
{"obj":"btn","id":4,"x":5,"y":5,"w":230,"h":58,"text":"开灯","value_ofs_x":-85,"value_font":32,"value_str":"\uE6E8","text_font":"notosans24"}
{"obj":"btn","id":5,"x":5,"y":68,"w":230,"h":58,"text":"日光","value_ofs_x":-85,"value_font":32,"value_str":"\uE599","text_font":"notosans24"}
{"obj":"btn","id":6,"x":5,"y":131,"w":230,"h":58,"text":"夜晚","value_ofs_x":-85,"value_font":32,"value_str":"\uE594","text_font":"notosans24"}
{"obj":"btn","id":7,"x":5,"y":194,"w":230,"h":58,"text":"关灯","value_ofs_x":-85,"value_font":32,"value_str":"\uE335","text_font":"notosans24"}
```
![image](https://user-images.githubusercontent.com/15969459/153293717-8db7989a-5b57-4298-917e-75401e4e03d4.png)

[notosans_chinese_3500.zip](https://github.com/HASwitchPlate/openHASP/files/8036425/notosans_chinese_3500.zip)

**Reply by fvanroie** [reply]
@erwin Will this be a solution for your question?

**Reply by erwin** [reply]
Just wanted to expand on this a little bit... I've done some testing with the Korean language, and it's also working great! Even easier than it was with Chinese.

Step 1: [Download the Korean Noto Sans Font from Google Fonts](https://fonts.google.com/noto/specimen/Noto+Sans+KR)

Step 2: Use @fvanroie's excellent [Online Font Tool](https://openhasp.haswitchplate.com/0.6.3/design/fonts/#online-font-converter)

In my case I used:

> Name: kor_20
> Size: 20
> Bpp: 3
> Source Font: NotoSansKR-Regular.ttf
> Ranges & Code Points: <<empty>>
> Symbols & Characters:

```
가각간갇갈갉갊감갑값갓갔강갖갗같갚갛개객갠갤갬갭갯갰갱갸갹걋걍걔걘걜거걱건걷걸걺검겁것겄헝겋게겐겔겜겝겟겠겡겨격겪견결겸겹겻겼경곁계곈곌곕계고곡곤곧골곪곬곯곰곱곳공곶과곽관괄괌괍괏광쾌괜쾰퀑괴괵괸괼굄굅굇굉교굔굘굡굣구국군굳굴굵굶굻굼굽굿궁궂궈권궐궤궤귀귄귈귐귑귓규균귤그극근귿글긁금급긋긍긔기긱긴긷길김깁깃깅깆깊까깍깎깐깔깖깜깝깟깠깡깥깨깩깬깰깸깹깻깼깽꺄꺅꺈꺼꺽꺾껀껄껌껍껏껐껑께껙껜껨껩껴견결겸겹겻겼경계고곡곤곧골곰곱곳공과곽관괄광쾌괴괵굄교구국군굴궁궈권귀규균귤그극근글금급긋긍기긴길김깁깃까깍깐깔깜깝깡깨깬깰깽꺼꺽껀껄껌껍껑께껴견결경계고곡곤곧골곰곱곳공과곽관괄광쾌괴괵굄교구국군굴궁궈권귀규균귤그극근글금급긋긍기긴길김깁깃나낙낚난날낡낢남납낫났낭낮낯낱낳내낵낸낼냄냅냇냈냉냐냑냔냥너넉넋넌널넒넓넘넙넛넜넝네넥넨넬넴넵넷넸넹녀녁년념렵녔녕녘녜노녹논놀놂놈놉놋농높놓놔놘놜놨뇌뇐뇔뇜뇝뇟뇨뇩뇬뇰뇸뇹뇽누눅눈눋눌눔눕눗눙눠눴눼뉘뉜뉠뉨뉩뉴늅늉느늑는늘늙늠늡늣능늦늪늬니닉닌닐님닙닛닝다닥닦단닫달닭닮닯닳담답닷닺닻닿대댁댄댈댐댑댓댔댕댜더덕던덜덺덤덥덧덩덫덮데덱덴델뎀뎁뎃뎄뎅뎌뎐뎔뎠령뎨도독돈돋돌돔돕돗동돛돝돠돤돨돼됐되된될됨됩됫됴두둑둔둘둠둡둣둥둬뒀뒈뒤뒨뒬뒵뒷듀듄듈듐듕드득든듣들듬듭듯등듸디딕딘딛딜딤딥딧딩딪따딱딴딸땀땁땃땄땅땋때땍땐땔땜땝땟땠땡떠떡떤떨떪떫떰떱떳떴떵떻떼떽뗀뗄뗌뗍뗏뗐뗑뗘뗬또똑똔똘똥똬똴뙈뙜뙤뙨뚜뚝뚠뚤뚫뚱뚸뛔뛰뛴뛸뜀뜁뜅뜨뜩뜬뜯뜰뜸뜹뜻띄띈띌띔띕띠띡띤띨띰띱띳띵라락란랄람랍랏랐랑랗래랙랜랠램랩랫랬랭랴략랸량러럭런럴럼럽럿렀렁렇게레렉렌렐렘렙렛렝려력련렬렴렵렷렸령례롄롑롓로록론롤롬롭롯롱롸롼뢍쾌뢴뢰뢸룀룁룃룅료룐룔룝룟룡루룩룬룰룸룹룻룽뤄뤘뤠뤼륀륄륌륏륭르륵른를름릅릇릉릐리릭린릴림립릿링마막만많맏말맑맒맘맙맛망맞맡맣매맥맨맬맴맵맷맸맹먀먁먈먕머먹먼멀멂멈멥멋멍메멕멘멜멤멥멧멨멩며멱면멸몃명몇몌모목몬몰몲몸몹못몽뫄뫈뫼묀묄묍묏묑묘묜묠묩묫무묵문묻물묽묾뭄뭅뭇뭉뭍뭐뭔뭘뭏뭐뮈뮌뮐뮤뮨뮬뮴뮷므믄믈름믐믑믓미믹민믿밀밈밉밋밍및밑바박밖밗반받발밝밞밟밤밥밧방밭배백밴밸뱀뱁뱃뱸뱅뱌뱍뱐뱡버벅번벌벎범법벗벙베벡벤벨벰벱벳벴벵벼벽변별볍볏볐병볕볘보복본볼봄봅봇봉봐봔봤뵤뵨부북분붇불붉붊붐붑붓붕붙뷔뷕뷘뷜뷩뷰뷴뷸븀븃븅브븍븐블븜븝븟비빅빈빌빔빕빗빙빚빛빠빡빤빨빪빰빱빳빺빵빻빼빽뺀뺄뺌뺍뺏뺐뺑뺘뺙뺨뻐뻑뻔뻘뻠뻣뻤뻥뻬뼁뼈뼉뼘뼙뼛뼜뼝뽀뽁뽄뽈뽐뽑뽕뾰뿅뿌뿍뿐뿔뿜뿡뿨뿩쀼쁑쁘쁜쁠쁨삐삑삔삘삠삡삣삥사삭산삳살삵삶삼삽삿샀상샂샃샆샇새색샌샐샘샘샙샛샜생샤샥샨샬샴샵샷샹섀섄섈섐섕서석섞선설섬섭섯섰성섶세섹센셀셈셉셋셌셍셔셕션셜셤셥셧셨셩셰셴셸솅소속손솔솖솜솝솟송솥솨솩솬솰솽쇄쇈쉘쇗쇘쇠쇤쇠쇨쇰쇱쇳쇼쇽숀숄숌숍숏숑수숙순숟술숨숩숫숭숯숱숲숴쉈쉐쉑쉔쉘쉠쉥쉬쉭쉰쉴쉼쉽쉿슁슈슉슐슘슛슝스슥슨슬슭슴습슷승시식신싣실싫심십싯싱싶싸싹싼쌀쌈쌉쌌쌍쌓쌔쌕쌘쌜쌤쌥쌨쌩썌써썩썬썰썲썸썹썼썽쎄쎈쎌쏀쎵쎼쏘쏙쏜쏟쏠쏢쏨쏩쏭쏴쏵쐈쐐쐬쐰쐴쐼쐽쑈쑤쑥쑨쑬쑴쑵쑹쒀쒔쒜쒸쒼쓔쓰쓱쓴쓸쓺씀씁씌씨씩씬씰씸씹씻씽아악안앉않알앍앎앓암압앗았앙앝앞애액앤앨앰앱앱앳앴앵야약얀얄얇얌얍얏양얕얗얘얜얠얩어억언얹얻얼얽얾엄업없엇었엉엊엌엎에엑엔엘엠엡엣엥여역엮연열엶염엽엾엿였영옅옆옇예옌옐옘옙옛옜오옥온올옭옮옯옰옳옴옵옷옹옻와왁완왈왐왑왓왔왕왜왝왠왬왜외왼욀욈욉외욋욍요욕욘욜욤욥욧용우욱운울욹욺움웁웃웅워웍원월웜웝웠웡웨웩웬웰웸웹웽위윅윈윌윔윕윗윙유육윤율윰윱윳융윷으윽은을읊음읍읏응읒읓읔읕읖읗의읜읠읨읫이익인일읽읾임입잇있잉잊잎자작잔잖잗잘잚잠잡잣잤장잦재잭잰챌잼잽잿쟀쟁쟈쟉쟌쟎쟐쟘쟝쟤쟨쟬저적전절젊점접젓정젖제젝젠젤젬젭젯젰젱져젼졀졌졍졔조족존졸졺좀좁좃종좆좇좋좌좍좔좝좟좡좨좼죄죈죌죔죕죗죙죠죤죵주죽준줄줆줌줍줏중줘줬줴쥐쥔쥘쥠쥡쥣쥬쥰쥴쥼즈즉즌즐즘즙즛증지직진짇질짊짐집짓징짖짙짚짜짝짠짤짧짬짭짯짰짱째짹짼쨀쨈쨽쨋쨌쨍쨔쨘쨩쩌쩍쩐쩔쩜쩝쩟쩠쩡쩨쩽쪄쪘쪼쪽쫀쫄쫌쫍쫏쫑쫓쫘쫙쫬쫴쬐쬔쬘쬠쬡쬬쭈쭉쭌쭐쭘쭙쭝쭤쭸쭹쮜쮸쯔쯤쯧쯩찌찍찐찔찜찝찡찢차착찬찮찰참찹찻찼창찾채책챈챌챔챕챗챘챙챠착챤챦챵처척천철첨첩첫첬청체첵첸첼쳄쳅쳇쳈쳉쳐쳔쳤쳬초촉촌촐촘숍촛총촤촨촬촹최쵠쵤쵬쵭쵯쵱쵸춈추축춘출춤춥춧충춰췄췌췐취췬췰췸췹췻췽츄츈츌츔층치칙친칟칠칡침칩칫칭카칵칸칼캄캅캇캉캐캑캔캘캠캡캣캤캥캬캭컁커컥컨컫컬컴컵컷컸컹케켁켄켈켐켑켓켕켜켠켤겸켯켰켱켸코콕콘콜콤콥콧콩콰콱콴괄쾅쾌쾐퀘퀑쾨쾰쿄쿠쿡쿤쿨쿰쿱쿳쿵쿼퀀퀄퀑퀘퀭퀴퀵퀸퀼큄큅큇큉큐큔큘큠크큭큰클큼큽큿킁키킥킨킬킴킵킷킹타탁탄탈탉탐탑탓탔탕태택탠탤탬탭탯탰탱탸턍터턱턴털턺텀텁텃텄텅테텍텐텔템텝텟텼텡텨텬텼톄토톡톤톨톰톱톳통톺톼퇀퇘퇴퇸퇼툄툅툇툉툐투툭툰툴툼툽툿퉁퉈퉜퉤튀튁튄튈튐튑튕튜튠튤튬튱트특튼튿틀틂틈틉틋틔틘틜틤틥티틱틴틸팀팁팃팅파팍팎판팔팖팜팝팟팠팡팥패팩팬팰팸팹팻팼팽퍄퍅퍼퍽펀펄펌펍펏펐펑페펙펜펠펨펩펫펭펴편펼폄폅폈평폐퓐퓔퓜포폭폰폴폼폽폳퐁퐈퐜푀푄표푠표푭표푱푸푹푼푿풀풂품풉풋풍풔풩퓌퓐퓔퓜퓟퓨퓬퓰퓸퓻퓽프픈플픔픕픗피픽핀필핌핍핏핑하학한할핥함합핫항해핵핸핼햄햅햇했행햐향허헉헌헐헒험헙헛헝헤헥헨헬헴헵헷헹혀혁현혈혐협혓혔형혜혠혤혭호혹혼홀홂홈홉홋홍홑화확환활황홰홱홴횃회획횐횔횜횝횟횡효횬횰횹횻횽후훅훈훌훑훔훕훗훙훠훤훨훵훼훽휀휄휑휘휵휜휠휨휩휫휭휴휵휸휼흄흇흉흐흑흔흖흗흘흙흠흡흣흥흩희흰흴흼흽힁히힉힌힐힘힙힛힝
0123456789
 
ABCDEFGHIJKLMNOPQRSTUVWXYZ
abcdefghijklmnopqrstuvwxyz
!@#$%^&*()-_+=|\
[]{}
:;"'
<>,.?/
```

NOTE: Do not delete the space - you definitely want that in your font.

And here's a preview of what it looks like... I've still got to fine tune it to get it all looking perfect, but it works!

<img width="419" height="565" alt="image" src="https://github.com/user-attachments/assets/cbb1e355-30a6-435f-91aa-c9708b9fb885" />

**Comment by kydang789** [comment]
Thanks you. I try

**Comment by fvanroie** [comment]
I have included an [online Font Converter tool](https://openhasp.haswitchplate.com/0.6.3/design/fonts/#custom-fonts) into the new documentation for the `0.6.3-dev` firmware.
You can use it to create customized binary fonts that can be uploaded to the flash partition.

Feel free to try it out and provide feedback here.

**Reply by kydang789** [reply]
Use offline converter is ok. But some time, I must reduce to 2 bitperpixel due to over firmware flash file??. But finally perfect. Convert to binary file, I not yet tried

**Reply by fvanroie** [reply]
What ESP32 board are you using? What is the total flash size?

**Reply by kydang789** [reply]
WT32-SC01 16MB. I try change option for 16MB flash again. I also bought you coffee. Thanks so much
https://www.buymeacoffee.com/aktdCofU/c/2616883

**Reply by erwin** [reply]
@fvanroie Could you elaborate on how to update `encodings.json` to support Chinese Simplified?

https://openhasp.haswitchplate.com/0.6.3/design/fonts/#character-sets

I created `chinese_20.bin` and `chinese_24.bin` via awesome new online [font convertor tool](https://openhasp.haswitchplate.com/0.6.3/design/fonts/#custom-fonts), using [Noto Sans Simplified Chinese](https://fonts.google.com/noto/specimen/Noto+Sans+SC) as the base font, and using the `2500` most common characters above, but I haven't updated my `encodings.json` and it doesn't seem to work until I do that.

Yet I don't quite understand how the `encodings.json` matches with the fonts specified in `encodings.json`... 

1. Do I just add a new `"chinese-simplified":` section to `encodings.json`?
2. Do I also need to generate `src/font/chinese-simplified/notosans-regular_{}_chinese_simplified.cpp`? And if so, any tips on that?
3. textfont = `src/font/NotoSansSC-Regular.otf`?
4. iconfont = `materialdesignicons-webfont.ttf`?
5. `chars`, any tips on how to bulk convert these? I saw in the documentation concerning formatting you wrote:
  * Code points up to 0xFFFF should be encoded as "\uE6E8" in the template.
    Note: Use lowercase \u and exactly 4 hexadecimal digits.

  * Code points above 0xFFFF must be encoded as "\U0001F5E9" in the template.
    Note: Use capital \U and exactly 8 hexadecimal digits.

  * Do I have to give every character a name? Can you think of an easy way to automate converting the list above into a range? The Unicode block of `4E00-9FFF` is supposed to cover all of the CJK common characters, but it's 20,991 characters!

From just scanning through the cjk-unified-ideographs in order, it seems likely that every character will need to be listed individually, as 4E00-9FFF is a very large range, and I think characters are organized by radical (sub character) rather than by frequency of use.

https://unicode-table.com/en/#cjk-unified-ideographs

**Reply by fvanroie** [reply]
@erwin You don't need to modify `encodings.json`. It is only used to generate the **internal** .cpp fonts using `tools\hasp_font_create.py`.

Since the Chinese characters are so numerous, I don't think it will be feasible to provide a firmware with Chinese default fonts.
It will simply be too big...

Therefor we now support the **external** fonts which can be loaded into PSram instead:

- Copy `chinese_20.bin` and `chinese_24.bin` onto the flash partition
- Then use `"text_font":"chinese_24"` and `"text":"关灯"` in the jsonl file

You need `0.6.3-dev` firmware to use the external binary fonts.

**Reply by erwin** [reply]
Thank you. I was just missing the upgrade to v0.6.3-dev firmware... I rebuilt the latest from master and it works great!

So cool!!!!

![image](https://user-images.githubusercontent.com/11534217/156156948-73421440-25b1-4b6d-8774-63421328880e.png)
